crates/client/contract-class-data/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+where
+    P: TransactionPool<Block = B>,
+    B: BlockT,
+{
+    receiver: UnboundedReceiver<DeclareTransactionStatusStream<P>>,
+    buffer: Vec<DeclareTransactionStatusStream<P>>,
+}
+
+impl<B, P> DeclareTransactionJobReceiverFuture<B, P>
+where
+    P: TransactionPool<Block = B>,
+    B: BlockT,
+{
+    #[inline(always)]
+    pub fn new(receiver: UnboundedReceiver<DeclareTransactionStatusStream<P>>, limit: NonZeroUsize) -> Self {
+        Self { receiver, buffer: Vec::with_capacity(limit.into()) }
+    }
+
+    #[inline(always)]
+    pub fn buffer_as_mut_ref(&mut self) -> &mut Vec<DeclareTransactionStatusStream<P>> {
+        &mut self.buffer
+    }
+}
+
+// When calling `poll_next`, if new values have been pushed to the channel, stores up to `limit` of
+// them in the buffer, and returns the exact amount.
+//
+// This whole struct only makes sense if limit > 0.
+// We use `buffer.capacity()` as limit, so  use the `new` method
+// that won't let you pass a non allocated `Vec`.
+impl<B, P> Stream for DeclareTransactionJobReceiverFuture<B, P>
+where
+    P: TransactionPool<Block = B>,
+    B: BlockT,
+{
+    type Item = usize;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let DeclareTransactionJobReceiverFuture { receiver, buffer, .. } = &mut *self;
+
+        match receiver.poll_recv_many(cx, buffer, buffer.capacity()) {
+            Poll::Pending => Poll::Pending,
+            // For limit > 0, this means that the channel has been closed
+            Poll::Ready(0) => Poll::Ready(None),
+            Poll::Ready(amount) => Poll::Ready(Some(amount)),
+        }
+    }
+}
+
+pub async fn run_worker<B, P>(
+    contract_class_db: Arc<ContractClassDataDb>,
+    mapping_db: Arc<MappingDb<B>>,
+    job_recv: UnboundedReceiver<DeclareTransactionStatusStream<P>>,
+) where
+    B: BlockT,
+    P: TransactionPool<Block = B>,
+{
+    // We push everything we get from the receiver here, then repetitively iterate over it to poll
+    // each transaction's status stream and take action when they end up being finalized.
+    let mut unfinalized_declare_transactions = Vec::new();
+    // Stores all the txs that were not yet mapped into our db when they finalized for later processing.
+    let mut finalized_but_not_mapped_yet = HashMap::<B::Hash, Vec<(TransactionHash, ClassHash)>>::new();
+    // Totally arbitrary limit.
+    // The job_recv being unbounded, there is not risk of it being full.
+    // Could be increased up to at most the max amount of declare tx possible to have in a single block.
+    // More that this would be useless as the "poll receiver" task is woke up at least once once but
+    // most likely multiple times) during the span of one blocktime.
+    let streams_buffer_limit = unsafe { NonZeroUsize::new_unchecked(10) };
+    let mut jobs_receiver = DeclareTransactionJobReceiverFuture::<B, P>::new(job_recv, streams_buffer_limit);
+
+    loop {
```

Comment:
or what we can do is change the `read_contract_class_data` so that if it doesn't find it in the persisted storages, it goes to look for it in the other one.

Reply:
Better, always store in the persisted storage, and just stores the keys of pending in the pending one

Reply:
Now there is no need to wait

Reply:
Ah the issue, is that you now return a value for a TX that may have failed...
Did not solve anything...

Reply:
Solution: we remove them from PENDING during the mapping process.

Reply:
done